Rework some preprocessing functions, test jobdb implementation for preprocessing

mmccrackan opened the pull request:
Sort of an update to #1156.  Does the following:

1. Switches to per-group parallelization for `preprocess_tod` and `multilayer_preprocess_tod`.
2. Reworks `preprocess_tod` and `multilayer_preprocess_tod` to rely more on `preproc_or_load_group` since it does many of the same things.
3. Start adding jobdb support for both `preprocess_tod` and `multilayer_preprocess_tod`.
4. Adds some standardized error messages in a class which can be input into a preprocessing error db.
5. Some more cleanup of functions and docstrings in `preprocess_util`.

Might split into multiple PRs if possible.  Will need to adjust some things to make sure it doesn't break things downstream.

msilvafe commented on sotodlib/preprocess/preprocess_util.py:
```diff
@@ -871,301 +967,299 @@ def cleanup_obs(obs_id, policy_dir, errlog, configs, context=None,
+                aman.move("tags", None)
+            aman.wrap('tags', tags_proc)
+
+            out_dict_proc = get_preproc_group_out_dict(obs_id, configs_proc,
```
In get_groups you dropped context_init/context_proc input and set to configs_init['context_file'] I'd do that consistently everywhere including here.

msilvafe commented on sotodlib/preprocess/preprocess_util.py:
```diff
@@ -1207,25 +1299,25 @@ def cleanup_mandb(error, outputs, configs, logger=None, overwrite=False):
-    elif error == 'load_success':
+    elif (
+        errors[0] == PreprocessErrors.LoadSuccess or
+        (errors[0] is None and out_dict is None)
```
When does this second case occur where errors[0] is None and out_dict is None?

mmccrackan commented on sotodlib/site_pipeline/multilayer_preprocess_tod.py:
```diff
@@ -10,231 +10,269 @@
+                j.mark_visited()
+                if errors[0] is not None:
+                    j.jstate = "failed"
+                    j.tags["error"] = errors[0]
```
The error message, and traceback are still being written to the error log in `cleanup_mandb` which should be triggered since a non-`None` error should be returned in all failure cases.

mmccrackan commented on sotodlib/site_pipeline/multilayer_preprocess_tod.py:
```diff
@@ -10,231 +10,269 @@
+        futures.append(executor.submit(pp_util.get_groups, obs["obs_id"], configs_proc))
+        futures_dict[futures[-1]] = obs
+
+    for future in tqdm(as_completed_callable(futures), total=len(futures),
```
Added this.  I can't seem to get it to overwrite the lines instead of appending, but I don't think that should be a major concern.  I've set it so it will only update for each percent so at most 100 lines.

mmccrackan commented:
Okay, after addressing comments and running some tests, I think this is probably in a state where it can be reviewed more officially.  I'll do some more testing of the various logic branches to look for failure cases but I've run the first 25 obs_ids from ISO v3 for SATp3 here:

`/global/cfs/cdirs/sobs/users/mmccrack/so/preprocess/common/20250821_preproc_jobdb/archives`

I've made some edits to the mapmaker to accomodate these changes, but these are untested.

We need to check and update the sims versions.

For future to-dos I'd like to update the error class to actually have its own exceptions but will do that in a different PR.  I was also thinking of removing error logging from the standard output to clean that up a bit and only putting it in the error log which should be much more parsable now.  I'll also do the merging of load_and_preprocess and multilayer_load_and_preprocess in a different PR after this.
